OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs

```diff
@@ -0,0 +1,357 @@
+	{
+		public CPos GetRandomBaseCenter()
+		{
+			var randomConstructionYard = World.Actors.Where(a => a.Owner == Player &&
+				Info.ConstructionYardTypes.Contains(a.Info.Name))
+				.RandomOrDefault(World.LocalRandom);
+
+			return randomConstructionYard != null ? randomConstructionYard.Location : initialBaseCenter;
+		}
+
+		public readonly World World;
+		public readonly Player Player;
+
+		readonly Func<Actor, bool> isEnemyUnit;
+		readonly Predicate<Actor> unitCannotBeOrdered;
+
+		public List<Squad> Squads = new List<Squad>();
+
+		IBotPositionsUpdated[] notifyPositionsUpdated;
+		IBotIdleBaseUnits[] notifyIdleBaseUnitCount;
+
+		CPos initialBaseCenter;
+		List<Actor> unitsHangingAroundTheBase = new List<Actor>();
+
+		// Units that the bot already knows about. Any unit not on this list needs to be given a role.
+		List<Actor> activeUnits = new List<Actor>();
+
+		int rushTicks;
+		int assignRolesTicks;
+		int attackForceTicks;
+		int minAttackForceDelayTicks;
+
+		public SquadManagerBotModule(Actor self, SquadManagerBotModuleInfo info)
+			: base(info)
+		{
+			World = self.World;
+			Player = self.Owner;
+
+			if (World.Type == WorldType.Editor)
+				return;
+
+			isEnemyUnit = unit =>
+				Player.Stances[unit.Owner] == Stance.Enemy
+					&& !unit.Info.HasTraitInfo<HuskInfo>()
+					&& unit.Info.HasTraitInfo<ITargetableInfo>();
+
+			unitCannotBeOrdered = a => a.Owner != Player || a.IsDead || !a.IsInWorld;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			notifyPositionsUpdated = Player.PlayerActor.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			notifyIdleBaseUnitCount = Player.PlayerActor.TraitsImplementing<IBotIdleBaseUnits>().ToArray();
+
+			// Avoid all AIs trying to rush in the same tick, randomize their initial rush a little.
+			var smallFractionOfRushInterval = Info.RushInterval / 20;
+			rushTicks = World.LocalRandom.Next(Info.RushInterval - smallFractionOfRushInterval, Info.RushInterval + smallFractionOfRushInterval);
+
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			assignRolesTicks = World.LocalRandom.Next(0, Info.AssignRolesInterval);
+			attackForceTicks = World.LocalRandom.Next(0, Info.AttackForceInterval);
+			minAttackForceDelayTicks = World.LocalRandom.Next(0, Info.MinimumAttackForceDelay);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			AssignRolesToIdleUnits(bot);
+		}
+
+		internal Actor FindClosestEnemy(WPos pos)
+		{
+			return World.Actors.Where(isEnemyUnit).ClosestTo(pos);
+		}
+
+		internal Actor FindClosestEnemy(WPos pos, WDist radius)
+		{
+			return World.FindActorsInCircle(pos, radius).Where(isEnemyUnit).ClosestTo(pos);
+		}
+
+		void CleanSquads()
+		{
+			Squads.RemoveAll(s => !s.IsValid);
+			foreach (var s in Squads)
+				s.Units.RemoveAll(unitCannotBeOrdered);
+		}
+
+		// Use of this function requires that there is one squad of this type. Hence it is a piece of shit
```

Comment:
Style nit: Can we reword this to `// HACK: Use of this function requires that there is one squad of this type.`?